docs/architecture.md

```diff
@@ -0,0 +1,222 @@
+
+### Schema
+
+```sql
+-- Table describes queue of verified CARs to be stored in filecoin
+-- CAR is considered in queue when there is no piece referencing it
+CREATE TABLE content
+(
+  -- CAR CID
+  link TEXT PRIMARY KEY,
+  -- CAR Size
+  size number NOT NULL,
+  -- Source where the content can be fetched from
+  source JSONB NOT NULL,
+  -- Timestamp
+  inserted TIMESTAMP WITH TIME ZONE DEFAULT timezone('utc'::text, now()) NOT NULL
+);
+
+-- Table describes pieces derived corresponding to CARs in the content table. Link (commP) is
+-- unique even though cargo reference is not, that is because there may be an error in piece
+-- derivation and in that case another correct piece will reference the same content.
+CREATE TABLE piece
+(
+  -- Piece CID
+  link TEXT NOT NULL PRIMARY KEY,
+  -- Piece size
+  size number NOT NULL,
+  -- Reference to the content of this piece (CAR CID).
+  content TEXT NOT NULL REFERENCES content(link),
+  -- Time when the piece was derived from the content
+  inserted TIMESTAMP WITH TIME ZONE DEFAULT timezone('utc'::text, now()) NOT NULL
+);
+
+CREATE INDEX piece_content_idx ON piece (content);
+CREATE INDEX piece_inserted_idx ON piece (inserted);
+
+-- Content for which we need to derive piece CIDs. We will have a process that
+-- reads from this queue and writes into `piece` table.
+CREATE VIEW content_queue AS
+  SELECT content.*
+  FROM content
+  LEFT OUTER JOIN piece ON content.link = piece.content
+  WHERE piece.content IS NULL
+  ORDER BY piece.inserted;
+
+-- Table describing pieces to be included into aggregates. If aggregate is NULL then the
+-- piece is queued for the aggregation.
+CREATE TABLE inclusion
+(
+  -- Piece CID. Notice that it is not unique because in case of bad piece
+  -- aggregate will be rejected and good pieces will be written back here
+  -- to be included into new aggregate
+  piece TEXT NOT NULL REFERENCES piece(link),
+  
+  -- Aggregate CID, if NULL the the piece is queued for the aggregation
+  aggregate TEXT REFERENCES aggregate(link) NULL,
+
+  -- Priority in the queue. I think it could be a counter on initial inserts hence
+  --  providing FIFO order. When piece is retried (after aggregate is rejected)
+  -- we could keep original priority hence prioritizing it over new items yet
+  -- keeping original FIFO order among rejects.
+  priority TEXT NOT NULL,
+
+  -- Time when the piece was added to the queue.
+  inserted TIMESTAMP WITH TIME ZONE DEFAULT timezone('utc'::text, now()) NOT NULL,
+
+  -- Piece may end up in multiple aggregates e.g. if aggregate was rejected
+  PRIMARY KEY (aggregate, piece)
+
+   -- We may also want to write inclusion proof here
+);
+
+CREATE INDEX inclusion_inserted_idx ON inclusion (inserted);
```

Comment:
Not sure why we need this given that we'll be sorting by `priority`, should we index that instead ?

Reply:
included priority